File: backend/app/services/crawler.py
import asyncio
import logging
from typing import Callable, Awaitable

from app.models.page import PageData
from app.services.confluence import ConfluenceClient

logger = logging.getLogger(__name__)


class PageTreeCrawler:
    """Recursively crawls Confluence page trees."""

    def __init__(
        self,
        client: ConfluenceClient,
        max_depth: int = -1,
        max_concurrent: int = 5,
        on_page_found: Callable[[str, int], Awaitable[None]] | None = None,
    ):
        self.client = client
        self.max_depth = max_depth
        self.max_concurrent = max_concurrent
        self.on_page_found = on_page_found
        self._semaphore = asyncio.Semaphore(max_concurrent)
        self.total_pages_found = 0

    async def crawl(self, page_id: str, depth: int = 0) -> PageData:
        """
        Crawl a page and all its children recursively.
        Returns a PageData tree.
        """
        async with self._semaphore:
            logger.info(f"Crawling page {page_id} (depth={depth})")
            page_data = await self._fetch_page(page_id, depth)
            self.total_pages_found += 1

            if self.on_page_found:
                await self.on_page_found(page_data.title, self.total_pages_found)

        # Check depth limit
        if self.max_depth != -1 and depth >= self.max_depth:
            return page_data

        # Crawl children concurrently
        child_tasks = []
        async for child_raw in self.client.iter_all_child_pages(page_id):
            child_id = child_raw["id"]
            task = asyncio.create_task(self.crawl(child_id, depth + 1))
            child_tasks.append(task)

        if child_tasks:
            children = await asyncio.gather(*child_tasks, return_exceptions=True)
            for child in children:
                if isinstance(child, PageData):
                    child.parent_id = page_id
                    page_data.children.append(child)
                elif isinstance(child, Exception):
                    logger.error(f"Failed to crawl child of {page_id}: {child}")

        return page_data

    async def _fetch_page(self, page_id: str, depth: int) -> PageData:
        """Fetch a single page and convert to PageData."""
        raw = await self.client.get_page(page_id)

        body_html = raw.get("body", {}).get("storage", {}).get("value", "")
        space_key = raw.get("space", {}).get("key", "")
        title = raw.get("title", "Untitled")

        # Build the page URL
        base_link = raw.get("_links", {}).get("base", "")
        web_link = raw.get("_links", {}).get("webui", "")
        url = base_link + web_link if base_link and web_link else ""

        return PageData(
            page_id=str(raw["id"]),
            title=title,
            space_key=space_key,
            body_html=body_html,
            url=url,
            depth=depth,
        )


def count_pages(page: PageData) -> int:
    """Count total pages in a page tree."""
    count = 1
    for child in page.children:
        count += count_pages(child)
    return count


def flatten_pages(page: PageData) -> list[PageData]:
    """Flatten a page tree into a list (BFS order)."""
    pages = [page]
    for child in page.children:
        pages.extend(flatten_pages(child))
    return pages

File: electron/src/services/crawler.ts
/**
 * Page-tree crawler — ported from backend/app/services/crawler.py.
 *
 * Uses a simple semaphore (max concurrent fetches) and recursively walks
 * /child/page until either max depth is hit or the tree is exhausted.
 */

import type { PageData } from '../types';
import { ConfluenceClient } from './confluence';

class Semaphore {
  private active = 0;
  private readonly waiters: Array<() => void> = [];
  constructor(private readonly limit: number) {}

  async acquire(): Promise<void> {
    if (this.active < this.limit) {
      this.active += 1;
      return;
    }
    await new Promise<void>((resolve) => this.waiters.push(resolve));
    this.active += 1;
  }

  release(): void {
    this.active -= 1;
    const next = this.waiters.shift();
    if (next) next();
  }
}

export type OnPageFound = (title: string, total: number) => void | Promise<void>;

export class PageTreeCrawler {
  private readonly semaphore: Semaphore;
  totalPagesFound = 0;

  constructor(
    private readonly client: ConfluenceClient,
    private readonly maxDepth: number = -1,
    maxConcurrent: number = 5,
    private readonly onPageFound?: OnPageFound,
  ) {
    this.semaphore = new Semaphore(maxConcurrent);
  }

  async crawl(pageId: string, depth = 0): Promise<PageData> {
    let pageData: PageData;
    await this.semaphore.acquire();
    try {
      console.info(`[crawler] Crawling page ${pageId} (depth=${depth})`);
      pageData = await this.fetchPage(pageId, depth);
      this.totalPagesFound += 1;
      if (this.onPageFound) {
        await this.onPageFound(pageData.title, this.totalPagesFound);
      }
    } finally {
      this.semaphore.release();
    }

    if (this.maxDepth !== -1 && depth >= this.maxDepth) {
      return pageData;
    }

    const childTasks: Array<Promise<PageData | Error>> = [];
    for await (const childRaw of this.client.iterAllChildPages(pageId)) {
      const childId = String(childRaw.id);
      childTasks.push(
        this.crawl(childId, depth + 1).catch((e: Error) => e),
      );
    }

    if (childTasks.length) {
      const results = await Promise.all(childTasks);
      for (const res of results) {
        if (res instanceof Error) {
          console.error(`[crawler] Failed to crawl child of ${pageId}:`, res);
        } else {
          res.parentId = pageId;
          pageData.children.push(res);
        }
      }
    }
    return pageData;
  }

  private async fetchPage(pageId: string, depth: number): Promise<PageData> {
    const raw = await this.client.getPage(pageId);
    const bodyHtml: string = raw?.body?.storage?.value ?? '';
    const spaceKey: string = raw?.space?.key ?? '';
    const title: string = raw?.title ?? 'Untitled';
    const baseLink: string = raw?._links?.base ?? '';
    const webLink: string = raw?._links?.webui ?? '';
    const url = baseLink && webLink ? baseLink + webLink : '';

    return {
      pageId: String(raw.id),
      title,
      spaceKey,
      bodyHtml,
      url,
      depth,
      children: [],
    };
  }
}

export function countPages(page: PageData): number {
  let count = 1;
  for (const child of page.children) count += countPages(child);
  return count;
}

export function flattenPages(page: PageData): PageData[] {
  const out = [page];
  for (const child of page.children) out.push(...flattenPages(child));
  return out;
}
